### ISSUE TITLE
Problems with file based databases (genome db cosmic, genome db tgi cancer-annotation, genome db tgi misc-annotation)

### ISSUE THREAD

**Issue by malachig** [genome/gms#56]
There are various issues with getting these installed in the standalone gms.  Here are some notes on these:

The HCC1395 test analysis needs three file based databases.  Try the following from within TGI (using the clin-seq demonstration model):
% genome model input show 2891454740
- cosmic/61.1
- tgi/cancer-annotation/human/build37-20130401.1
- tgi/misc-annotation/human/build37-20130113.1

These file based databases are actually git repos.  They exist inside TGI but some of them have also been uploaded to github to either 'genome' or 'genome-vendor'.  For example:

https://github.com/genome-vendor/genome-db-tgi-cancer-annotation
https://github.com/genome-vendor/genome-db-tgi-misc-annotation
https://github.com/genome-vendor/genome-db-cosmic-data

Code relevant to these file based databases:
Genome/Db/*
Genome/Db/Command/InstallFromGitRepo.pm
Genome/Db/Cosmic/Command/Install.pm
Genome/Db/Tgi/Command/Install.pm

**Comment by malachig** [comment]
The first problem is that these repos are assumed in Genome/Db/Command/InstallFromGitRepo.pm to live in 'genome-vendor' but two of them are in 'genome'.

Another problem is that the latest COSMIC exceeds github file size limits.

Scott proposed a primitive solution as follows:
1. Use plain old unix "split" on all of the big files (split FILE PREFIX) split SOMEBIGFILE SOMEBIGFILE.gitsplit.
2. Have a script in the repo called "pre-commit" that does this on all of the big files.
3. Have a script in the repo called "post-update" the re-combines them with "cat".

**Comment by malachig** [comment]
From Scott:

To make a new version, it's easiest to create it in the final location TGI expects it to live:

ssh deploy
cd /gsc/scripts/opt/genome/db/cosmic

#We expect a directory for each version number, with a checkout at that version
git clone git@github.com:genome-vendor/genome-db-cosmic-data.git $THE_NEW_VERSION
#alternatively, cp one of the other directories for speed
cp -a $THE_LATEST_OLD_VERSION $THE_NEW_VERSION

#get on the old branch just in case
git checkout $THE_LATEST_OLD_VERSION

#make a branch for the new version
git checkout -b $THE_NEW_VERSION

#dump the new files in place <<<<

#commit things
git add *
git commit -m "new version $THE_NEW_VERSION"
git push origin $THE_NEW_VERSION

#now it should be installable!
#but you will probably want to test before commit/push
#at least do this:
genome db list --filter source_name=cosmic

#if you do work in a non-standard location, you can switch $GENOME_DB environment dir:
mkdir ~/mydb/cosmic/
mv $THE_REPO ~/mydb/cosmic/$THE_NEW_VERSION
GENOME_DB=$HOME/mydb/ genome db list --source_name=cosmic
GENOME_DB=$HOME/mydb somecommand

**Comment by malachig** [comment]
To rename/move repos once they are in github:

To move a repo, you change the "owner" of the repo.
- Go to the main page for the repo.
- Click the Settings button on the right.
- Scroll to the bottom.
- Choose "Transfer ownership".
- e.g. switch the owner to "genome-vendor".

There is also a box to rename the repository in the setting page

When you change the repo owner or name, you'll want to change the "remote" for the local git checkouts:

ssh deploy
cd /gsc/scripts/opt/genome/db/cosmic/$VERSION
git remote -v
git remote rename origin old
git remote add origin $NEW_URI
git fetch origin 
git pull origin -u $NEW_VERSION # reset the default remote
git remote rm old

**Comment by malachig** [comment]
Another issue with these DBs is that they are expected to have a symlink 'latest' within the top dir of each repo pointing to the 'latest' branch.  

When for example, 'genome db cosmic install' runs, it does not seem to create this symlink correctly.  If this symlink is missing 'genome db list' will not work and possibly other things as well.

**Comment by malachig** [comment]
When the user goes to install one of these databases, if they wind up attempting to clone a branch that does not exist, old versions of git warn but then clone the default branch instead of dying.

We can explicitly check that the correct branch was cloned by doing something like this:

bash -c 'cd $DIR; git checkout $BRANCH'
echo $?

**Comment by malachig** [comment]
I have reworked the code in 'Genome/Db/Command/InstallFromGitRepo.pm' to be more fault tolerant, more explicit in how the commands are structured and to create what appears to be correctly installed databases for all three sources using the following commands.
- genome db cosmic install --branch=65_v2  (id: cosmic/65.2)
- genome db tgi install --subsource=cancer-annotation --species=human --branch=human-build37-20130113 (id: tgi/cancer-annotation/human/human-build37-20130113.1)
- genome db tgi install --subsource=misc-annotation --species=human --branch=human-build37-20130113 (id: tgi/misc-annotation/human/human-build37-20130113.1)

I committed this to gms-core: master and gms-pub branches
https://github.com/genome/gms-core/commit/f5b123527eed6d9c517edde6acb0ad68ec4e6dea
https://github.com/genome/gms-core/commit/f6eb256354120dfe207acbfc3f9a3855ba24d543

However, to test properly we need the right versions of each database to be up on github. These are
- cosmic/61.1
- tgi/cancer-annotation/human/build37-20130401.1
- tgi/misc-annotation/human/build37-20130113.1

In other words, we need to get these commands working:
- genome db cosmic install --branch=61_v1 (install working!)
- genome db tgi install --subsource=cancer-annotation --species=human --branch=human-build37-20130401 (install working!)
- genome db tgi install --subsource=misc-annotation --species=human --branch=human-build37-20130113 (install working!)

We also still need to determine a way to deal with large files that github does not allow...

Once all of that is working we will need to update 'Genome/Model/Command/Import/Metadata.pm' to display the correct install commands in the form above.

**Comment by malachig** [comment]
Note that this issue needs to be resolved to fix issues #13 and #50

**Comment by malachig** [comment]
My last round of fixes to 'Genome/Db/Command/InstallFromGitRepo.pm' were not quite there.  The Db name for databases installed in a standalone environment were not quite matching the conventions used within TGI.  There is code in gms-core that is quite particular about how these are set up.  I have applied a further tweak to get them to come out right.  This fix was tested within the TGI and then committed to both master and gms-pub branches:

https://github.com/genome/gms-core/commit/60f42d892c3f858a1bcac711a7ea5ac665911101 (master)
https://github.com/genome/gms-core/commit/4de4a8d42279edc20e0a1babf67968937aa611b9 (gms-pub)

I then pulled this change into clia1 where I will test as well.

**Comment by sakoht** [comment]
If you relocate things at TGI I doubt anyone is attached to the old way.  I previously put symlinks in place to keep old snapshots working until they aged out.

**Comment by gatoravi** [comment]
I've made similar split files for cosmic 61.1 as cosmic 65.2 and tried to push these files into a new branch. But looks like I don't have access to push to genome-vendor.

The files were split using,
`split -a 2 -d -l 250000  cosmic.NCBI-human.ensembl.70_37_v5.anno ''`

The original files can be reconstructed with a simple 
`cat *`

**Comment by sakoht** [comment]
Weird.  I'll see if I can change that.  If not @malachig or @obigriffith or @jasonwalker80 or @nnutter can change it...

**Comment by sakoht** [comment]
It would be great if the split/merge happen in such a standard way that it could be done automatically when installing.

**Comment by sakoht** [comment]
Actually the list of people with access is now really short.

cc: @acoffman @ebelter @jweible and @tmooney have access currently and can add you.

Sent from my iPhone

**Comment by gatoravi** [comment]
Version 61.1 of Cosmic is now on Github(genome-vendor) as a separate branch. Thanks to @nnutter for setting this up. This has to be tested on the standalone installs with the new `genome db install` code.

**Comment by malachig** [comment]
I completed yet another set of tweaks to 'genome db install' here:
https://github.com/genome/gms-core/commit/36c50a326fdf5a6a5d15ee0251d2df9bd9406ce9
https://github.com/genome/gms-core/commit/da2dd78e5f95f470259c5c0f533f0288053f6440

The correct versions of all three databases can now be installed successfully on clia1 using the following commands:
- genome db cosmic install --branch=61_v1
- genome db tgi install --subsource=cancer-annotation --species=human --branch=human-build37-20130401
- genome db tgi install --subsource=misc-annotation --species=human --branch=human-build37-20130113

The next steps are to: 
(1) Automate splitting/merging of large files in the cosmic db 
(2) Automatically run these db installations during 'genome model import metadata' so that the end user does not have to think about any of this. 
(3) Update the tutorial to remove the current example db install commands.  The current note that these are only needed by clin-seq is false.  Also they really do not take very long to install.  We should just do this automatically to simplify the flow for users attempting the HCC1395 demonstration analysis.

**Comment by malachig** [comment]
I believe all the remaining issues described above have been resolved by the following:

https://github.com/genome-vendor/genome-db-cosmic-data/blob/61_v1/join_db.pl
https://github.com/genome-vendor/genome-db-cosmic-data/blob/61_v1/split_db.pl
https://github.com/genome/gms/commit/b2fa6e4ba9010ae7a5060ba7291709942a34573f
https://github.com/genome/gms-core/commit/e4839a601ff5af13f7f98832380cbf7fffacb895
https://github.com/genome/gms-core/commit/4aba8d8d9dc0576a4dc8930178e2f687ea613d87

Correct installation of the appropriate file based databases from github should now happen automatically when the user runs 'genome model import metadata'

As a final step, we should make sure that everything is working as expected for steps in somatic-variation and clin-seq pipelines where these databases are actually used.

**Comment by malachig** [comment]
I previously forgot to commit and push changes to 'genome model import metadata' so that the correct 'genome db $source install' commands are run.  This is now done and committed to gms-core master and gms-pub branches here:

https://github.com/genome/gms-core/commit/d92670aa034dc9d5f938627ba22c6370b61e219d
https://github.com/genome/gms-core/commit/c8318573a12d78d7fc3b459065b08fbda5bd9717

**Comment by malachig** [comment]
This seems to all be working smoothly now.  We are routinely installing cosmic, tgi/misc-annotation, and tgi/cancer-annotation into sGMS instances.  It really doesn't take that long so we just do it automatically during the metadata import stage.  I think that errors related to these databases in somatic-variation and rna-seq have been resolved.  There may still be problems in clin-seq though.  Leaving this issue open until we have made it through there.

**Comment by malachig** [comment]
Clin-Seq/Med-Seq is mostly working now and the few issues remaining do not seem to be related to the file-based databases.  Closing for now.
